File: eda/04_seasonal_extremes.py
"""
eda/04_seasonal_extremes.py
Analyze summer extremes: zero-output events, extended low-output periods,
and patterns that could indicate extreme heat derating.
"""
import pandas as pd
import numpy as np
from pathlib import Path
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
from matplotlib.dates import DateFormatter

from table_utils import write_table

SCRIPT_STEM = "04_seasonal_extremes"
TRACES = Path("data/2024/pisp-downloads/Traces")
FIGURES = Path("eda/figures")
FIGURES.mkdir(parents=True, exist_ok=True)

HH_COLS_SOL = [str(i) for i in range(1, 49)]
HH_COLS_WIND = [str(i).zfill(2) for i in range(1, 49)]
MIDDAY = [str(i) for i in range(24, 36)]

def load_trace(tech, year, location):
    f = TRACES / f"{tech}_{year}" / f"{location}_RefYear{year}.csv"
    if not f.exists():
        return None
    df = pd.read_csv(f)
    df['datetime'] = pd.to_datetime(
        df['Year'].astype(int).astype(str) + '-' +
        df['Month'].astype(int).astype(str).str.zfill(2) + '-' +
        df['Day'].astype(int).astype(str).str.zfill(2)
    )
    return df

# ====== 1. Find hottest historical summers and examine solar patterns ======
# Known hot Australian summers:
# 2019: Black Summer (bushfires, extreme heat)
# 2013: "Angry Summer" (record heat)
# 2017: Hot summer in NSW/VIC
# 2015: Strong El Niño

HOT_SUMMERS = [2019, 2013, 2017, 2015, 2023]
COOL_SUMMERS = [2011, 2016, 2020, 2022]  # La Niña years

SOLAR_LOC = 'Bannerton_SAT'
WIND_LOC = 'DUNDWF1'

# ====== Figure 1: Hot vs Cool summer solar profiles ======
fig, axes = plt.subplots(2, 1, figsize=(14, 10))

hot_cool_summer_rows = []
for ax, season_type, year_list, color in [
    (axes[0], 'Hot Summers', HOT_SUMMERS, 'darkred'),
    (axes[1], 'Cool Summers', COOL_SUMMERS, 'steelblue'),
]:
    for yr in year_list:
        df = load_trace('solar', yr, SOLAR_LOC)
        if df is None:
            continue
        summer = df[df['Month'].isin([12, 1, 2])]
        if len(summer) == 0:
            continue
        daily = summer[HH_COLS_SOL].mean(axis=1)
        hot_cool_summer_rows.append({
            "season_type": season_type,
            "year": yr,
            "n_days": len(daily),
            "mean_daily_cf": daily.mean(),
            "std_daily_cf": daily.std(),
            "min_daily_cf": daily.min(),
            "max_daily_cf": daily.max(),
        })
        ax.plot(summer['datetime'], daily, linewidth=0.5, alpha=0.6,
               color=color, label=str(yr) if yr == year_list[0] else None)
        ax.plot(summer['datetime'], daily.rolling(3).mean(), linewidth=1.5,
               color='black', alpha=0.8)

    ax.set_title(f"Solar {SOLAR_LOC} — {season_type} Daily Mean CF")
    ax.set_ylabel("Daily Mean CF")
    ax.set_ylim(0, 0.5)
    ax.grid(True, alpha=0.3)
    ax.legend(loc='upper right', fontsize=8)

axes[1].set_xlabel("Date")
plt.tight_layout()
plt.savefig(FIGURES / "04_hot_vs_cool_summer_solar.png", dpi=120, bbox_inches='tight')
plt.close()
print(f"Saved: 04_hot_vs_cool_summer_solar.png")

hot_cool_path = write_table(pd.DataFrame(hot_cool_summer_rows), SCRIPT_STEM, "hot_cool_summer_solar_summary")
print(f"Saved table: {hot_cool_path}")

# ====== Figure 2: Extended low-output events (3+ consecutive days) ======
fig2, axes2 = plt.subplots(2, 2, figsize=(14, 10))

combined_low_events = []
for ax, tech, loc, hh_cols in [
    (axes2[0, 0], 'solar', SOLAR_LOC, HH_COLS_SOL),
    (axes2[0, 1], 'wind', WIND_LOC, HH_COLS_WIND),
]:
    # Find multi-day low-output events across all years
    all_low_events = []
    for yr in range(2011, 2024):
        df = load_trace(tech, yr, loc)
        if df is None:
            continue
        summer = df[df['Month'].isin([12, 1, 2])]
        if len(summer) == 0:
            continue
        daily = summer[hh_cols].mean(axis=1)
        threshold = 0.1 if tech == 'solar' else 0.15

        # Find consecutive days below threshold
        below = (daily < threshold).astype(int)
        diff = below.diff()
        starts = diff[diff == 1].index
        ends = diff[diff == -1].index

        for s, e in zip(starts, ends):
            duration = (e - s) + 1
            if duration >= 3:
                all_low_events.append({
                    'year': yr,
                    'start': summer.loc[s, 'datetime'],
                    'end': summer.loc[e, 'datetime'],
                    'duration': duration,
                    'min_cf': daily.loc[s:e].min(),
                    'mean_cf': daily.loc[s:e].mean(),
                    'tech': tech,
                })

    combined_low_events.extend(all_low_events)

    events_df = pd.DataFrame(all_low_events)
    if len(events_df) > 0:
        # Histogram of event duration
        ax.hist(events_df['duration'], bins=range(1, max(events_df['duration']) + 2),
               color='darkorange' if tech == 'solar' else 'steelblue', alpha=0.7, edgecolor='black')
        ax.set_title(f"{tech.upper()} {loc} — Duration of Low-Output Events (≥3 days)")
        ax.set_xlabel("Consecutive Days Below Threshold")
        ax.set_ylabel("Count")
        ax.grid(True, alpha=0.3)

        # Print worst events
        worst = events_df.nsmallest(5, 'mean_cf')
        print(f"\n{tech.upper()} worst summer low-output events:")
        for _, row in worst.iterrows():
            print(f"  {row['year']}: {row['start'].date()} to {row['end'].date()} "
                  f"({row['duration']}d, mean={row['mean_cf']:.3f}, min={row['min_cf']:.3f})")
    else:
        ax.text(0.5, 0.5, 'No multi-day low-output events found',
               ha='center', va='center', transform=ax.transAxes)
        ax.set_title(f"{tech.upper()} {loc} — No Extended Low Events")

low_output_events_path = write_table(pd.DataFrame(combined_low_events), SCRIPT_STEM, "low_output_events")
print(f"Saved table: {low_output_events_path}")

# ====== Figure 3: Worst single-day solar events ======
ax3 = axes2[1, 0]
worst_solar_days = []
for yr in range(2011, 2024):
    df = load_trace('solar', yr, SOLAR_LOC)
    if df is None:
        continue
    summer = df[df['Month'].isin([12, 1, 2])]
    if len(summer) == 0:
        continue
    daily = summer[HH_COLS_SOL].mean(axis=1)
    worst_idx = daily.idxmin()
    worst_solar_days.append({
        'year': yr,
        'date': summer.loc[worst_idx, 'datetime'],
        'cf': daily.loc[worst_idx],
    })

worst_df = pd.DataFrame(worst_solar_days).sort_values('cf')

worst_solar_day_rows = [
    {
        "year": int(row['year']),
        "date": row['date'].strftime("%Y-%m-%d"),
        "cf": row['cf'],
        "is_hot_summer": 1 if row['year'] in HOT_SUMMERS else 0,
    }
    for _, row in worst_df.iterrows()
]
worst_solar_day_path = write_table(pd.DataFrame(worst_solar_day_rows), SCRIPT_STEM, "worst_solar_day_summary")
print(f"Saved table: {worst_solar_day_path}")

colors = ['darkred' if y in HOT_SUMMERS else 'steelblue' for y in worst_df['year']]
ax3.bar(range(len(worst_df)), worst_df['cf'], color=colors, alpha=0.7)
ax3.set_xticks(range(len(worst_df)))
ax3.set_xticklabels([f"{r['year']}" for _, r in worst_df.iterrows()], rotation=45, fontsize=8)
ax3.set_title(f"Solar {SOLAR_LOC} — Worst Summer Day by Year (sorted)")
ax3.set_ylabel("Daily Mean CF")
ax3.grid(True, alpha=0.3)

# ====== Figure 4: Half-hourly profile of worst solar day ======
ax4 = axes2[1, 1]
worst_solar_day_profile_rows = []
if len(worst_df) > 0:
    worst_row = worst_df.iloc[0]
    yr = int(worst_row['year'])
    worst_date = worst_row['date']
    df = load_trace('solar', yr, SOLAR_LOC)
    if df is not None:
        mask = (df['datetime'] == worst_date)
        if mask.any():
            worst_day = df[mask]
            half_hours = np.arange(0.5, 24.5, 0.5)
            ax4.plot(half_hours, worst_day.iloc[0, 3:51], 'darkred', linewidth=2, marker='o', markersize=3)
            ax4.set_title(f"Worst Solar Day: {worst_date.date()} (CF={worst_row['cf']:.3f})")
            ax4.set_xlabel("Hour")
            ax4.set_ylabel("Capacity Factor")
            ax4.set_ylim(0, 1)
            ax4.grid(True, alpha=0.3)
            for hh, cf in zip(half_hours, worst_day.iloc[0, 3:51]):
                worst_solar_day_profile_rows.append({
                    "year": yr,
                    "date": worst_date.strftime("%Y-%m-%d"),
                    "half_hour": hh,
                    "cf": cf,
                })

plt.tight_layout()
plt.savefig(FIGURES / "04_low_output_events.png", dpi=120, bbox_inches='tight')
plt.close()
print(f"Saved: 04_low_output_events.png")

worst_solar_day_profile_path = write_table(
    pd.DataFrame(worst_solar_day_profile_rows), SCRIPT_STEM, "worst_solar_day_profile"
)
print(f"Saved table: {worst_solar_day_profile_path}")

# ====== Figure 5: Hourly CF distribution by month (solar) ======
fig5, ax5 = plt.subplots(figsize=(12, 6))

df_all = load_trace('solar', 2019, SOLAR_LOC)  # Use 2019 as representative hot year
if df_all is not None:
    df_all['month'] = df_all['Month']
    monthly_stats = df_all.groupby('month')[HH_COLS_SOL].agg(['mean', 'std', 'min', 'max'])

    months = range(1, 13)
    means = [monthly_stats.loc[m].xs('mean', level=1).mean() if m in monthly_stats.index else 0 for m in months]
    stds = [monthly_stats.loc[m].xs('std', level=1).mean() if m in monthly_stats.index else 0 for m in months]

    ax5.bar(months, means, yerr=stds, color='darkorange', alpha=0.6, edgecolor='black')
    ax5.set_title(f"Solar {SOLAR_LOC} 2019 — Monthly Mean CF ± Std")
    ax5.set_xlabel("Month")
    ax5.set_ylabel("Mean Capacity Factor")
    ax5.set_xticks(months)
    ax5.set_xticklabels(['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
                        'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'])
    ax5.grid(True, alpha=0.3)

plt.tight_layout()
plt.savefig(FIGURES / "04_monthly_cf_2019.png", dpi=120, bbox_inches='tight')
plt.close()
print(f"Saved: 04_monthly_cf_2019.png")

monthly_cf_rows = [
    {"month": m, "mean_cf": means[i], "std_cf": stds[i]}
    for i, m in enumerate(months)
]
monthly_cf_path = write_table(pd.DataFrame(monthly_cf_rows), SCRIPT_STEM, "monthly_cf_2019_summary")
print(f"Saved table: {monthly_cf_path}")

# ====== Figure 6: Summer 2019 detailed — Black Summer ======
fig6, ax6 = plt.subplots(figsize=(16, 5))
df_2019 = load_trace('solar', 2019, SOLAR_LOC)
if df_2019 is not None:
    summer_2019 = df_2019[df_2019['Month'].isin([12, 1, 2])]
    daily = summer_2019[HH_COLS_SOL].mean(axis=1)
    rolling3_2019 = daily.rolling(3).mean()
    ax6.plot(summer_2019['datetime'], daily, linewidth=0.5, color='darkorange', alpha=0.7)
    ax6.plot(summer_2019['datetime'], rolling3_2019, linewidth=2, color='darkred')
    ax6.set_title(f"Solar {SOLAR_LOC} — Summer 2019 (Black Summer)")
    ax6.set_ylabel("Daily Mean CF")
    ax6.set_ylim(0, 0.5)
    ax6.grid(True, alpha=0.3)
    ax6.xaxis.set_major_formatter(DateFormatter('%Y-%m-%d'))
    plt.xticks(rotation=45)

plt.tight_layout()
plt.savefig(FIGURES / "04_summer_2019_black_summer.png", dpi=120, bbox_inches='tight')
plt.close()
print(f"Saved: 04_summer_2019_black_summer.png")

if df_2019 is not None:
    black_summer_rows = [
        {
            "date": d.strftime("%Y-%m-%d"),
            "daily_mean_cf": daily_cf,
            "rolling3_cf": rolling3_cf,
        }
        for d, daily_cf, rolling3_cf in zip(summer_2019['datetime'], daily, rolling3_2019)
    ]
    black_summer_path = write_table(pd.DataFrame(black_summer_rows), SCRIPT_STEM, "black_summer_2019_daily_cf")
    print(f"Saved table: {black_summer_path}")

print("\nDone.")

File: eda/04_seasonal_extremes.jl
#!/usr/bin/env julia

using CSV
using DataFrames
using Dates
using Printf
using Statistics

const SCRIPT_STEM = "04_seasonal_extremes"
const TRACES = joinpath("data", "2024", "pisp-downloads", "Traces")
const TABLE_ROOT = joinpath(@__DIR__, "tables")

function table_dir(script_stem; producer = "julia", root = TABLE_ROOT)
    path = joinpath(root, producer, script_stem)
    mkpath(path)
    return path
end

function table_path(script_stem, table_name; producer = "julia", root = TABLE_ROOT)
    filename = endswith(table_name, ".csv") ? table_name : "$(table_name).csv"
    return joinpath(table_dir(script_stem; producer = producer, root = root), filename)
end

function write_table(frame::DataFrame, script_stem, table_name; producer = "julia", root = TABLE_ROOT)
    path = table_path(script_stem, table_name; producer = producer, root = root)
    CSV.write(path, frame; missingstring = "")
    println("Saved table: ", path)
    return path
end

const HH_COLS_SOL = string.(1:48)
const HH_COLS_WIND = [lpad(i, 2, '0') for i in 1:48]

# Known hot vs cool (La Nina) Australian historical summers.
const HOT_SUMMERS = [2019, 2013, 2017, 2015, 2023]
const COOL_SUMMERS = [2011, 2016, 2020, 2022]

const SOLAR_LOC = "Bannerton_SAT"
const WIND_LOC = "DUNDWF1"

function add_datetime!(df::DataFrame)
    df.datetime = Date.(df.Year, df.Month, df.Day)
    return df
end

function load_trace(tech, yr, loc)
    file = joinpath(TRACES, "$(tech)_$(yr)", "$(loc)_RefYear$(yr).csv")
    isfile(file) || return nothing
    df = CSV.read(file, DataFrame)
    add_datetime!(df)
    return df
end

row_mean(df::DataFrame, cols) = [mean(row[col] for col in cols) for row in eachrow(df)]

# Rolling mean matching pandas' `Series.rolling(window).mean()` default
# `min_periods == window`: the first `window - 1` entries are missing.
function rolling_mean(values, window)
    n = length(values)
    result = Vector{Union{Missing, Float64}}(missing, n)
    for i in window:n
        result[i] = mean(values[(i - window + 1):i])
    end
    return result
end

# ====== Table 1: Hot vs cool summer solar profile summary ======
function write_hot_cool_summary_table()
    rows = NamedTuple[]
    for (season_type, year_list) in (("Hot Summers", HOT_SUMMERS), ("Cool Summers", COOL_SUMMERS))
        for yr in year_list
            df = load_trace("solar", yr, SOLAR_LOC)
            df === nothing && continue
            summer_mask = in.(df.Month, Ref((12, 1, 2)))
            any(summer_mask) || continue
            vals = row_mean(df[summer_mask, :], HH_COLS_SOL)
            push!(
                rows,
                (
                    season_type = season_type,
                    year = yr,
                    n_days = length(vals),
                    mean_daily_cf = mean(vals),
                    std_daily_cf = std(vals),
                    min_daily_cf = minimum(vals),
                    max_daily_cf = maximum(vals),
                ),
            )
        end
    end
    write_table(DataFrame(rows), SCRIPT_STEM, "hot_cool_summer_solar_summary")
end

# ====== Table 2: Multi-day (3+) low-output events across all years ======
#
# Mirrors pandas' `below.diff()` + `starts = diff[diff==1].index` /
# `ends = diff[diff==-1].index` + `zip(starts, ends)` exactly, including its
# quirks:
#   - `.diff()` is positional (state transitions in filtered-row order), but
#     `starts`/`ends` capture the ORIGINAL (unfiltered) row position labels,
#     which are not contiguous once rows are filtered to Dec/Jan/Feb only.
#   - `duration = (e - s) + 1` is then computed from those labels directly,
#     not from an actual count of matched rows, so an event whose state runs
#     from the tail of a Dec/Jan/Feb block into the next one (across the
#     excluded Mar-Nov gap) gets a hugely inflated (but faithfully
#     reproduced) duration.
#   - `zip(starts, ends)` pairs the two lists purely by position; if their
#     lengths differ (e.g. a run already active at the first row leaves an
#     unmatched trailing "end"), every subsequent pair is shifted and the
#     resulting durations can even go negative. This is replicated verbatim.
#   - `daily.loc[s:e]` is a label-based range select: all filtered rows whose
#     original position falls within [s, e], which is what `mask_range` below
#     reproduces.
function low_output_events_for(tech, loc, hh_cols, threshold, yr)
    df = load_trace(tech, yr, loc)
    df === nothing && return NamedTuple[]
    summer_mask = in.(df.Month, Ref((12, 1, 2)))
    any(summer_mask) || return NamedTuple[]

    full_idx = findall(summer_mask)  # original row positions, in file order
    dates = df.datetime[full_idx]
    daily = row_mean(df[full_idx, :], hh_cols)
    below = daily .< threshold
    n = length(below)

    starts = Int[]
    ends = Int[]
    for i in 2:n
        delta = Int(below[i]) - Int(below[i - 1])
        if delta == 1
            push!(starts, full_idx[i])
        elseif delta == -1
            push!(ends, full_idx[i])
        end
    end

    rows = NamedTuple[]
    for k in 1:min(length(starts), length(ends))
        s = starts[k]
        e = ends[k]
        duration = (e - s) + 1
        duration >= 3 || continue
        mask_range = (full_idx .>= s) .& (full_idx .<= e)
        vals = daily[mask_range]
        s_pos = findfirst(==(s), full_idx)
        e_pos = findfirst(==(e), full_idx)
        push!(
            rows,
            NamedTuple{(:year, :start, :end, :duration, :min_cf, :mean_cf, :tech)}(
                (
                    yr,
                    Dates.format(dates[s_pos], "yyyy-mm-dd"),
                    Dates.format(dates[e_pos], "yyyy-mm-dd"),
                    duration,
                    minimum(vals),
                    mean(vals),
                    tech,
                ),
            ),
        )
    end
    return rows
end

function write_low_output_events_table()
    rows = NamedTuple[]
    for (tech, loc, hh_cols, threshold) in (
        ("solar", SOLAR_LOC, HH_COLS_SOL, 0.1),
        ("wind", WIND_LOC, HH_COLS_WIND, 0.15),
    )
        for yr in 2011:2023
            append!(rows, low_output_events_for(tech, loc, hh_cols, threshold, yr))
        end
    end
    write_table(DataFrame(rows), SCRIPT_STEM, "low_output_events")
end

# ====== Table 3: Worst summer day by year (all years, solar) ======
function write_worst_solar_day_table()
    rows = NamedTuple[]
    for yr in 2011:2023
        df = load_trace("solar", yr, SOLAR_LOC)
        df === nothing && continue
        summer_mask = in.(df.Month, Ref((12, 1, 2)))
        any(summer_mask) || continue
        summer = df[summer_mask, :]
        daily = row_mean(summer, HH_COLS_SOL)
        worst_pos = argmin(daily)  # first occurrence on ties, matching pandas idxmin
        push!(
            rows,
            (
                year = yr,
                date = Dates.format(summer.datetime[worst_pos], "yyyy-mm-dd"),
                cf = daily[worst_pos],
                is_hot_summer = yr in HOT_SUMMERS ? 1 : 0,
            ),
        )
    end
    write_table(DataFrame(rows), SCRIPT_STEM, "worst_solar_day_summary")
    return rows
end

# ====== Table 4: Half-hourly profile of the single worst solar day (all years) ======
function write_worst_solar_day_profile_table(worst_rows)
    rows = NamedTuple[]
    if !isempty(worst_rows)
        best_idx = argmin([r.cf for r in worst_rows])
        worst = worst_rows[best_idx]
        yr = worst.year
        worst_date = Date(worst.date, "yyyy-mm-dd")
        df = load_trace("solar", yr, SOLAR_LOC)
        if df !== nothing
            mask = df.datetime .== worst_date
            if any(mask)
                row_idx = findfirst(mask)
                half_hours = collect(0.5:0.5:24.0)
                for (hh, col) in zip(half_hours, HH_COLS_SOL)
                    push!(rows, (year = yr, date = worst.date, half_hour = hh, cf = df[row_idx, col]))
                end
            end
        end
    end
    write_table(DataFrame(rows), SCRIPT_STEM, "worst_solar_day_profile")
end

# ====== Table 5: Monthly CF summary for 2019 ======
#
# Replicates the (corrected) two-stage aggregation: per half-hourly column,
# compute that column's mean/std across the days in the month, then average
# those 48 per-column values into one monthly figure. Matches the fixed
# Python `monthly_stats.loc[m].xs('mean'|'std', level=1).mean()` expression
# (the original `.loc[m, ('mean',)]` / `('mean', std)` lookups raised
# KeyError/NameError and were fixed as part of this port).
function write_monthly_cf_2019_table()
    df = load_trace("solar", 2019, SOLAR_LOC)
    rows = NamedTuple[]
    if df !== nothing
        for m in 1:12
            mask = df.Month .== m
            if any(mask)
                sub = df[mask, :]
                col_means = [mean(sub[!, col]) for col in HH_COLS_SOL]
                col_stds = [std(sub[!, col]) for col in HH_COLS_SOL]
                push!(rows, (month = m, mean_cf = mean(col_means), std_cf = mean(col_stds)))
            else
                push!(rows, (month = m, mean_cf = 0.0, std_cf = 0.0))
            end
        end
    end
    write_table(DataFrame(rows), SCRIPT_STEM, "monthly_cf_2019_summary")
end

# ====== Table 6: Black Summer 2019 detailed daily trace ======
function write_black_summer_table()
    df = load_trace("solar", 2019, SOLAR_LOC)
    rows = NamedTuple[]
    if df !== nothing
        summer_mask = in.(df.Month, Ref((12, 1, 2)))
        summer = df[summer_mask, :]
        daily = row_mean(summer, HH_COLS_SOL)
        rolling3 = rolling_mean(daily, 3)
        for (i, d) in enumerate(summer.datetime)
            push!(rows, (date = Dates.format(d, "yyyy-mm-dd"), daily_mean_cf = daily[i], rolling3_cf = rolling3[i]))
        end
    end
    write_table(DataFrame(rows), SCRIPT_STEM, "black_summer_2019_daily_cf")
end

function main()
    write_hot_cool_summary_table()
    write_low_output_events_table()
    worst_rows = write_worst_solar_day_table()
    write_worst_solar_day_profile_table(worst_rows)
    write_monthly_cf_2019_table()
    write_black_summer_table()

    println("\nDone.")
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
